src/app/+item-page/simple/entity-types/shared/entity-page-fields.component.ts

```diff
@@ -0,0 +1,120 @@
+import { Relationship } from '../../../../core/shared/entities/relationship.model';
+import { Item } from '../../../../core/shared/item.model';
+import { getRemoteDataPayload } from '../../../../core/shared/operators';
+import { hasValue } from '../../../../shared/empty.util';
+import { ITEM } from '../../../../shared/entities/switcher/entity-type-switcher.component';
+
+const compareArraysUsing = <T>(mapFn: (t: T) => any) =>
+  (a: T[], b: T[]): boolean => {
+    if (!Array.isArray(a) || ! Array.isArray(b)) {
+      return false
+    }
+
+    const aIds = a.map(mapFn);
+    const bIds = b.map(mapFn);
+
+    return aIds.length === bIds.length &&
+      aIds.every((e) => bIds.includes(e)) &&
+      bIds.every((e) => aIds.includes(e));
+  };
+
+const compareArraysUsingIds = <T extends { id: string }>() =>
+  compareArraysUsing((t: T) => hasValue(t) ? t.id : undefined);
+
+/**
+ * Fetch the relationships which match the type label given
+ * @param {string} label      Type label
+ * @returns {(source: Observable<[Relationship[] , RelationshipType[]]>) => Observable<Relationship[]>}
+ */
+export const filterRelationsByTypeLabel = (label: string) =>
+  (source: Observable<[Relationship[], RelationshipType[]]>): Observable<Relationship[]> =>
+    source.pipe(
+      map(([relsCurrentPage, relTypesCurrentPage]) =>
+        relsCurrentPage.filter((rel: Relationship, idx: number) =>
+          hasValue(relTypesCurrentPage[idx]) && (relTypesCurrentPage[idx].leftLabel === label ||
+          relTypesCurrentPage[idx].rightLabel === label)
+        )
+      ),
+      distinctUntilChanged(compareArraysUsingIds())
+    );
+
+/**
+ * Operator for turning a list of relationships into a list of the relevant items
+ * @param {string} thisId           The item's id of which the relations belong to
+ * @param {ItemDataService} ids     The ItemDataService to fetch items from the REST API
+ * @returns {(source: Observable<Relationship[]>) => Observable<Item[]>}
+ */
+export const relationsToItems = (thisId: string, ids: ItemDataService) =>
+  (source: Observable<Relationship[]>): Observable<Item[]> =>
+    source.pipe(
+      flatMap((rels: Relationship[]) =>
+        Observable.zip(
+          ...rels.map((rel: Relationship) => {
+            let queryId = rel.leftId;
+            if (rel.leftId === thisId) {
+              queryId = rel.rightId;
+            }
+            return ids.findById(queryId);
+          })
+        )
+      ),
+      map((arr: Array<RemoteData<Item>>) =>
+        arr
+          .filter((d: RemoteData<Item>) => d.hasSucceeded)
+          .map((d: RemoteData<Item>) => d.payload)),
+      distinctUntilChanged(compareArraysUsingIds()),
+    );
+
+@Component({
+  selector: 'ds-entity-page-fields',
+  template: ''
+})
+/**
+ * A generic component for displaying metadata and relations of an item
+ */
+export class EntityPageFieldsComponent implements OnInit {
```

Comment:
This is the superclass of all other PageFieldsComponents, it's never instantiated so an argument could be made that it should be abstract.

PageFieldsComponents are called that way because their templates determine which _fields_ are shown on the simple item _page_ of their entity-type.

For example: The [PublicationPageFieldsComponent](https://github.com/DSpace/dspace-angular/blob/ec6124035d35a505b6d6dec17305baa8220d3fd4/src/app/%2Bitem-page/simple/entity-types/publication/publication-page-fields.component.html) is the most like the regular Item Page in dspace. Its template contains which fields are shown on a Publication Page and how they're structured (in columns, the left one narrow, the right one wide, and both full width on small screens). On top of the regular metadata fields it also shows related People, Projects, Org Units and Journal Issues. These are configured in a similar way to a metadata field.

The [PersonPageFieldsComponent](https://github.com/DSpace/dspace-angular/blob/ec6124035d35a505b6d6dec17305baa8220d3fd4/src/app/+item-page/simple/entity-types/person/person-page-fields.component.html) does the same thing for items of the type Person. It specifies that for a person we don't want to see issue dates, abstracts and Journal Issues, but instead ORCiDs, Staff IDs and Job Titles.

The `item-page-*-field` components you mention render a single field of a certain type, e.g. an issue date, or an abstract. The PageFieldsComponents determine which of these fields should be shown on the page for a given entity-type.

Reply:
@artlowel : I think I mostly understand. However, would it be more accurate to just call this the `PublicationPageComponent` (i.e. it defines what appears on the Publication Page) ? Based on the [Component definition](https://github.com/DSpace/dspace-angular/blob/ec6124035d35a505b6d6dec17305baa8220d3fd4/src/app/%2Bitem-page/simple/entity-types/publication/publication-page-fields.component.ts) itself, it seems to be more than just metadata fields, as it also stores relationships.

Naming these all `*PageFieldsComponent`s still seems a bit awkward (to me), as we don't have a generic `ItemPageFieldsComponent` (for a generic item).  But, we *do* have an `ItemPageFieldComponent` (singular "field") which is named similarly but has a very different purpose of displaying just one metadata field.

I guess my point here is that (as a developer just looking at this), the names are confusing to me, so I'm wondering out loud if there's a way to make this clearer to new developers.

Reply:
I'd rather not call them *PageComponent, because everywhere else, the component that corresponds to the route is called the PageComponent. And these ones don't correspond to the route, that's what the ItemPageComponent does. Each of the *PageFieldsComponents is shown inside the ItemPageComponent.

But I agree that it's not the best name… How about simply (Publication|Journal|Person|…)Component? The version where it's shown as a part of a list is called *ListElementComponent, so if you see those two I think it will be clear which one renders the full page.

Reply:
@artlowel : Renaming these to `(Publication|Journal|Person|..)Component` is fine by me.  I think that's clearer than `PageFieldsComponent`.

Reply:
We've done that now